src/basic/devnum-util.c

```diff
@@ -57,6 +58,42 @@ int device_path_make_major_minor(mode_t mode, dev_t devnum, char **ret) {
         return 0;
 }
 
+int device_path_make_major_minor2(mode_t mode, dev_t devnum, char **ret) {
+        const char *t;
+        _cleanup_free_ char *syspath = NULL;
+        _cleanup_free_ char *link = NULL;
+        _cleanup_free_ char *fname = NULL;
+
+        /* Generates the /dev/... path given a dev_t */
+
+        if (S_ISCHR(mode))
+                t = "char";
+        else if (S_ISBLK(mode))
+                t = "block";
+        else
+                return -ENODEV;
+
+        if (asprintf(&syspath, "/sys/dev/%s/" DEVNUM_FORMAT_STR, t, DEVNUM_FORMAT_VAL(devnum)) < 0)
+                return -ENOMEM;
+
+        int r = readlink_malloc(syspath, &link);
+        if (r < 0)
+                return r;
+
+        r = path_extract_filename(link, &fname);
+        if (r < 0)
+                return r;
+
+        if (asprintf(ret, "/dev/%s", fname) < 0)
+                return -ENOMEM;
+
+        struct stat st;
+        if (stat(*ret, &st) < 0)
+                return -errno;
+
```

Comment:
the reason why i suggested to have the stat() here is to verify the assumption that the the device node is named like the basename of the sysfs path of the device. i.e. after you stat() here, please check:

```c
if (((mode ^ st.mode) & S_IFMT) != 0)
        return S_ISBLK(mode) ? -ENOTBLK : -ENODEV;
if (st.rdev != devnum)
        return -ENXIO;
```

this will check that device node type (i.e. char vs. block) and major/minor actually match our expectations.